web/src/engine/main/src/keymanEngine.ts

```diff
@@ -137,31 +137,57 @@ export default class KeymanEngine<
     });
 
     this.contextManager.on('keyboardchange', (kbd) => {
-      this.refreshModel();
-      this.core.activeKeyboard = kbd?.keyboard;
-
-      this.legacyAPIEvents.callEvent('keyboardchange', {
-        internalName: kbd?.metadata.id ?? '',
-        languageCode: kbd?.metadata.langId ?? ''
-      });
-
       // Hide OSK and do not update keyboard list if using internal keyboard (desktops).
       // Condition will not be met for touch form-factors; they force selection of a
       // default keyboard.
       if(!kbd) {
         this.osk.startHide(false);
       }
 
-      if(this.osk) {
-        this.osk.setNeedsLayout();
-        this.osk.activeKeyboard = kbd;
-        this.osk.present();
+      const earlyBatchClosure = () => {
+        this.refreshModel();
+        // Triggers context resets that can trigger layout stuff.
+        // It's not the final such context-reset, though.
+        this.core.activeKeyboard = kbd?.keyboard;
+
+        this.legacyAPIEvents.callEvent('keyboardchange', {
+          internalName: kbd?.metadata.id ?? '',
+          languageCode: kbd?.metadata.langId ?? ''
+        });
+      }
+
+      /*
+        Needed to ensure the correct layer is displayed AND that deadkeys from
+        the old keyboard have been wiped.
+
+        Needs to be after the OSK has loaded for the keyboard in case the default
+        layer should be something other than "default" for the current context.
+      */
+      const doContextReset = () => {
+        this.contextManager.resetContext();
       }
```

Comment:
Is there a reason this needs to be a closure? It's just one simple line called a few lines later in the function